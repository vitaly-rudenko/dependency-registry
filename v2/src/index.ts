const INVALID_CLASSES: unknown[] = [Object, Array, Number, String, Symbol, Function]

type LazyValue<T> = (dependencies: T) => T[keyof T]

export type Factory<T> = T extends { new (...args: infer C): infer I }
  ? (...args: C) => I
  : never

function isClass(value: unknown): value is ({ new (...args: unknown[]): unknown }) {
  const properties = Object.getOwnPropertyNames(value)
  return properties.includes('prototype') && !properties.includes('arguments')
}

function capitalize<T extends string>(input: T): Capitalize<T> {
  return input[0].toUpperCase() + input.slice(1) as Capitalize<T>
}

function validateName(name: unknown): asserts name is string {
  if (!name || typeof name !== 'string') {
    throw new Error(`Invalid name: ${String(name)}`)
  }
}

export class DependencyRegistry<T extends { [name: string]: unknown }> {
  private readonly dependencies = new Map<keyof T, T[keyof T] | LazyValue<T>>()
  private readonly lazyDependencies = new Set<keyof T>()
  // @ts-expect-error TODO
  private readonly proxy = new Proxy<T>({} as T, {
    ownKeys: () => {
      return [...this.dependencies.keys()] as string[]
    },
    has: (_, name) => {
      return this.dependencies.has(name as string)
    },
    getOwnPropertyDescriptor: (_, name) => {
      if (typeof name !== 'string' || !this.dependencies.has(name)) {
        return undefined
      }

      return {
        value: this.proxy[name],
        enumerable: true,
        configurable: true,
      }
    },
    get: (_, name) => {
      if (name === Symbol.iterator) {
        return () => {
          const keys = this.dependencies.entries()
          return { next: () => keys.next() }
        }
      }

      if (!name) {
        throw new Error('Unsupported action')
      }

      if (typeof name !== 'string') {
        throw new Error(`Unsupported action: ${String(name)}`)
      }

      const dependency = this.dependencies.get(name)
      if (dependency === undefined) {
        throw new Error(`Unknown dependency: "${name}"`)
      }

      if (this.isLazy(name, dependency)) {
        const value = dependency(this.export())
        if (value === undefined) {
          throw new Error(`Lazy value is undefined: ${name}`)
        }

        this.lazyDependencies.delete(name)
        this.dependencies.set(name, value)
        return value
      }

      return dependency
    }
  })

  value<Name extends Extract<keyof T, string>>(name: Name, value: T[Name]): void {
    validateName(name)

    if (value === undefined) {
      throw new Error(`Lazy value cannot be undefined`)
    }

    if (this.dependencies.has(name)) {
      throw new Error(`Value is already registered: ${String(name)}`)
    }

    this.dependencies.set(name, value)
  }

  lazy<Name extends Extract<keyof T, string>>(name: Name, value: (dependencies: T) => T[Name]): void {
    validateName(name)

    if (this.dependencies.has(name)) {
      throw new Error(`Lazy value is already registered: ${String(name)}`)
    }

    this.dependencies.set(name, value)
    this.lazyDependencies.add(name)
  }

  factory<
    Name extends { [K in keyof T]: K extends `create${infer N}` ? Uncapitalize<N> : never }[keyof T],
    FactoryName extends keyof T = `create${Capitalize<Name>}`
  >(
    name: Name,
    value: T[FactoryName] extends (...args: infer C) => infer I
      ? { new (dependencies: T, ...args: C): I } | ((dependencies: T, ...args: C) => I)
      : never,
  ): void {
    validateName(name)

    if (value === undefined) {
      throw new Error(`Factory value cannot be undefined`)
    }

    const factoryName = `create${capitalize(name)}` as FactoryName
    if (this.dependencies.has(factoryName)) {
      throw new Error(`Factory is already registered: ${String(name)}`)
    }

    if (isClass(value)) {
      if (INVALID_CLASSES.includes(value)) {
        throw new Error(`Invalid factory class: ${String(value.name)}`)
      }

      // @ts-expect-error TODO
      this.dependencies.set(factoryName, (...args: C) => new value(this.export(), ...args))
    } else if (typeof value === 'function') {
      // @ts-expect-error TODO
      this.dependencies.set(factoryName, (...args: C) => value(this.export(), ...args))
    } else {
      throw new Error(`Invalid factory value: ${String(name)}`)
    }
  }

  export(): T {
    return this.proxy as T
  }

  private isLazy<Name extends keyof T>(name: Name, value: unknown): value is LazyValue<T> {
    return this.lazyDependencies.has(name)
  }
}
