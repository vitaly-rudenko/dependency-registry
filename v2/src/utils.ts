export function validateName(name: unknown): asserts name is string {
  if (!name || typeof name !== 'string') {
    throw new Error(`Invalid name: '${String(name)}'`)
  }
}

export type DependencyName<Deps> = Extract<keyof Deps, string>
