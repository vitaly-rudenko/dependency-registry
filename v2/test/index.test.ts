/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

import { DependencyRegistry } from '../src'

const values = [
  123,
  true,
  'hello world',
  [1, 2, 3],
  { hello: 'world' },
  Symbol('hello world'),
  () => true,
  Buffer.from('hello world'),
]

const invalidNames = [
  undefined,
  null,
  '',
  true,
  Symbol('hello world'),
  123,
  { hello: 'world' },
  [1, 2, 3],
]

describe('DependencyRegistry', () => {
  let registry: DependencyRegistry<any>

  beforeEach(() => {
    registry = new DependencyRegistry()
  })

  describe('export()', () => {
    it('throws an error for unknown dependencies', () => {
      expect(() => registry.export().name).toThrowErrorMatchingInlineSnapshot(`"Unknown dependency: "name""`)
    })

    it('throws an error for unsupported actions', () => {
      expect(() => registry.export()['']).toThrowErrorMatchingInlineSnapshot(`"Unsupported action"`)
      expect(() => registry.export()[Symbol('name')]).toThrowErrorMatchingInlineSnapshot(`"Unsupported action: Symbol(name)"`)
      expect(() => registry.export()[Symbol.asyncIterator]).toThrowErrorMatchingInlineSnapshot(`"Unsupported action: Symbol(Symbol.asyncIterator)"`)
    })

    it('supports spread operator, iterators and "has" operator', () => {
      registry.value('firstName', 'John')
      registry.lazy('lastName', () => 'Doe')
      registry.factory('fullName', () => 'John Doe')

      const { firstName, ...rest } = registry.export()

      expect(firstName).toBe('John')
      expect(rest).toStrictEqual({ lastName: 'Doe', createFullName: expect.any(Function) })

      expect('firstName' in registry.export()).toBe(true)
      expect('lastName' in registry.export()).toBe(true)
      expect('createFullName' in registry.export()).toBe(true)
      expect('middleName' in registry.export()).toBe(false)

      expect(Object.entries(registry.export())).toStrictEqual([
        ['firstName', 'John'],
        ['lastName', 'Doe'],
        ['createFullName', expect.any(Function)],
      ])

      for (const item of registry.export()) {
        expect(item).toStrictEqual([expect.any(String), expect.anything()])
      }

      expect({...registry.export()}).toStrictEqual({
        firstName: 'John',
        lastName: 'Doe',
        createFullName: expect.any(Function),
      })

      expect([...registry.export()]).toStrictEqual([
        ['firstName', 'John'],
        ['lastName', 'Doe'],
        ['createFullName', expect.any(Function)],
      ])
    })

    it('allows plain export', () => {
      registry.value('hello', 'world')

      const proxy = registry.export()
      const plain = registry.export({ plain: true })

      expect(proxy).not.toBe(plain)

      expect(plain).toStrictEqual({ hello: 'world' })
      expect(plain.goodbye).toBeUndefined()
    })
  })

  describe('value()', () => {
    it.each(values)('registers a value', (value) => {
      registry.value('name', value)

      const { name } = registry.export()

      expect(name).toBe(value)
    })

    it('fails when value is already registered', () => {
      registry.value('hello', 'world')

      expect(() => registry.value('hello', 'there')).toThrowErrorMatchingInlineSnapshot(`"Value is already registered: hello"`)

      expect(registry.export().hello).toBe('world')
    })

    it.each(invalidNames)('fails when name is invalid', (name) => {
      expect(() => registry.value(name, 'value')).toThrowErrorMatchingSnapshot()
    })

    it('fails when value is invalid', () => {
      expect(() => registry.value('name', undefined)).toThrowErrorMatchingInlineSnapshot(`"Lazy value cannot be undefined"`)
    })
  })

  describe('lazy()', () => {
    it.each(values)('registers a lazy value', (value) => {
      registry.lazy('name', () => value)

      const { name } = registry.export()

      expect(name).toBe(value)
    })

    it('initializes lazy value once when accessed', () => {
      const spy = jest.fn(() => 'value')

      registry.lazy('name', () => spy())

      const exported = registry.export()

      expect(spy).not.toHaveBeenCalled()

      expect(exported.name).toBe('value')
      expect(exported.name).toBe('value')
      expect(exported.name).toBe('value')

      expect(spy).toHaveBeenCalledTimes(1)
    })

    it('initializes lazy value with dependencies', () => {
      registry.value('firstName', 'John')
      registry.lazy('lastName', () => 'Doe')
      registry.lazy('fullName', ({ firstName, lastName }) => `${firstName} ${lastName}`)

      expect(registry.export().fullName).toBe('John Doe')
    })

    it('fails when value is already registered', () => {
      registry.lazy('hello', () => 'world')

      expect(() => registry.lazy('hello', () => 'there')).toThrowErrorMatchingInlineSnapshot(`"Lazy value is already registered: hello"`)

      expect(registry.export().hello).toBe('world')
    })

    it.each(invalidNames)('fails when name is invalid', (name) => {
      expect(() => registry.lazy(name, () => 'value')).toThrowErrorMatchingSnapshot()
    })

    it('fails when value is invalid', () => {
      registry.lazy('name', () => undefined)

      expect(() => registry.export().name).toThrowErrorMatchingInlineSnapshot(`"Lazy value is undefined: name"`)
    })
  })

  describe('factory()', () => {
    it.each(values)('registers a factory', (value) => {
      registry.factory('name', () => value)

      const { createName } = registry.export()

      expect(createName()).toBe(value)
    })

    it('creates a factory from a class', () => {
      class Greeting {
        constructor(readonly dependencies, readonly lastName: string) {}
        generate() { return `Hello, ${this.dependencies.firstName} ${this.lastName}!` }
      }

      registry.value('firstName', 'John')
      registry.factory('greeting', Greeting)

      const { createGreeting } = registry.export()

      const greeting = createGreeting('Doe')

      expect(greeting).toBeInstanceOf(Greeting)
      expect(greeting.generate()).toBe('Hello, John Doe!')
    })

    it('fails when value is already registered', () => {
      registry.factory('hello', () => 'world')

      expect(() => registry.factory('hello', () => 'there')).toThrowErrorMatchingInlineSnapshot(`"Factory is already registered: hello"`)

      expect(registry.export().createHello()).toBe('world')
    })

    it.each(invalidNames)('fails when name is invalid', (name) => {
      expect(() => registry.factory(name, () => 'value')).toThrowErrorMatchingSnapshot()
    })

    it.each([
      undefined,
      123,
      true,
      [1, 2, 3],
      'hello world',
      { hello: 'world' },
      Buffer.from('hello world'),
    ])('fails when value is invalid', (value) => {
      expect(() => registry.factory('name', value)).toThrowErrorMatchingSnapshot()
    })

    it.each([
      Object, Array, Number, String, Symbol, Function,
    ])('fails when provided class is not valid', (value) => {
      expect(() => registry.factory('name', value)).toThrowErrorMatchingSnapshot()
    })
  })
})
