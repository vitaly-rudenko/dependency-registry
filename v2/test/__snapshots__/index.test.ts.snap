// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyRegistry factory() fails when name is invalid 1`] = `"Invalid factory name: undefined"`;

exports[`DependencyRegistry factory() fails when name is invalid 2`] = `"Invalid factory name: null"`;

exports[`DependencyRegistry factory() fails when name is invalid 3`] = `"Invalid factory name: "`;

exports[`DependencyRegistry factory() fails when name is invalid 4`] = `"Invalid factory name: true"`;

exports[`DependencyRegistry factory() fails when name is invalid 5`] = `"Invalid factory name: Symbol(hello world)"`;

exports[`DependencyRegistry factory() fails when name is invalid 6`] = `"Invalid factory name: 123"`;

exports[`DependencyRegistry factory() fails when name is invalid 7`] = `"Invalid factory name: [object Object]"`;

exports[`DependencyRegistry factory() fails when name is invalid 8`] = `"Invalid factory name: 1,2,3"`;

exports[`DependencyRegistry factory() fails when value is invalid 1`] = `"Factory value cannot be undefined"`;

exports[`DependencyRegistry factory() fails when value is invalid 2`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry factory() fails when value is invalid 3`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry factory() fails when value is invalid 4`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry factory() fails when value is invalid 5`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry factory() fails when value is invalid 6`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry factory() fails when value is invalid 7`] = `"Invalid factory value: name"`;

exports[`DependencyRegistry lazy() fails when name is invalid 1`] = `"Invalid name: undefined"`;

exports[`DependencyRegistry lazy() fails when name is invalid 2`] = `"Invalid name: null"`;

exports[`DependencyRegistry lazy() fails when name is invalid 3`] = `"Invalid name: "`;

exports[`DependencyRegistry lazy() fails when name is invalid 4`] = `"Invalid name: true"`;

exports[`DependencyRegistry lazy() fails when name is invalid 5`] = `"Invalid name: Symbol(hello world)"`;

exports[`DependencyRegistry lazy() fails when name is invalid 6`] = `"Invalid name: 123"`;

exports[`DependencyRegistry lazy() fails when name is invalid 7`] = `"Invalid name: [object Object]"`;

exports[`DependencyRegistry lazy() fails when name is invalid 8`] = `"Invalid name: 1,2,3"`;

exports[`DependencyRegistry value() fails when name is invalid 1`] = `"Invalid name: undefined"`;

exports[`DependencyRegistry value() fails when name is invalid 2`] = `"Invalid name: null"`;

exports[`DependencyRegistry value() fails when name is invalid 3`] = `"Invalid name: "`;

exports[`DependencyRegistry value() fails when name is invalid 4`] = `"Invalid name: true"`;

exports[`DependencyRegistry value() fails when name is invalid 5`] = `"Invalid name: Symbol(hello world)"`;

exports[`DependencyRegistry value() fails when name is invalid 6`] = `"Invalid name: 123"`;

exports[`DependencyRegistry value() fails when name is invalid 7`] = `"Invalid name: [object Object]"`;

exports[`DependencyRegistry value() fails when name is invalid 8`] = `"Invalid name: 1,2,3"`;
